lib/HLSL/LiveValueAnalysis.cpp

```diff
@@ -0,0 +1,1323 @@
+      // Only count the values for which metadata exists.
+      // TODO: Report count of values for those which are not used metadata?
+      // this can lead to missing live values in some cases.
+      if( isa<ExtractValueInst>( I ) ) {
+        I = dyn_cast<Instruction>(I->getOperand( 0 ));
+      }
+      if( I->isUsedByMetadata() ) {
+        regs++;
+      }
+      else {
+        DetectedInstr.push_back( I );
+      }
+      detected++;
+    }
+    if( detected != regs )
+    {
+      PrettyStr << "** DEBUG ************\n";
+      VSStr << "** DEBUG ************\n";
+      PrettyStr << "Detected " << (int)detected << " live values but only " << (int)regs << " are used by metadata\n";
+      VSStr << "Detected " << (int)detected << " live values but only " << (int)regs << " are used by metadata\n";
+      for( auto I : DetectedInstr ) {
+        I->print( PrettyStr );
+        I->print( VSStr );
+        PrettyStr << "\n";
+        VSStr << "\n";
+      }
+      PrettyStr << "*********************\n";
+      VSStr << "*********************\n";
+    }
+    if( const llvm::DebugLoc &debugInfo = callSite->getDebugLoc() ) {
+      PrettyStr << "Total 32-bit registers: " << (int)regs << "\n\n";
+      VSStr << "Total 32-bit registers: " << (int)regs << "\n\n";
+    }
+
+    std::ostringstream TmpStr;
+    for( Instruction *I :
+      sortInstructions<Instruction *>( m_spillsPerTraceCall[callSite] ) ) {
+      TmpStr.str( "" );
+      {
+        if( isa<ExtractValueInst>( I ) ) {
+          I = dyn_cast<Instruction>(I->getOperand( 0 ));
+        }
+
+        if( I->isUsedByMetadata() ) {
+          unsigned int totalLineWidth = 0;
+
+          if( auto *L = LocalAsMetadata::getIfExists( I ) ) {
+            if( auto *MDV = MetadataAsValue::getIfExists( I->getContext(), L ) ) {
+              for( User *U : MDV->users() ) {
+                if( DbgValueInst *DVI = dyn_cast<DbgValueInst>(U) ) {
+
+                  TmpStr << translateValueToName( DVI, I );
+
+                  // Output USE locations
+                  std::vector<PHINode *> Phis;
+                  std::set<PHINode *> VisitedPhis;
+                  std::set<DILocation *> UseLocations;
+                  while( 1 ) {
+                    for( User *U : I->users() ) {
+                      if( auto Inst = dyn_cast<Instruction>(U) ) {
+                        if( DILocation *Loc = Inst->getDebugLoc() ) {
+                          // Manually search for duplication location;
+                          // DILocations can be distinct but their source
+                          // location is not
+                          bool bFound = false;
+                          DIScope *scope = Loc->getScope();
+                          for( auto L : UseLocations ) {
+                            if( L->getScope() == scope ) {
+                              if( L->getLine() == Loc->getLine() &&
+                                L->getColumn() == Loc->getColumn() ) {
```

Comment:
might as well make them all part of the same if

Reply:
Changed.